oauth2_http/java/com/google/auth/oauth2/S2A.java

```diff
@@ -0,0 +1,110 @@
+package com.google.auth.oauth2;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.JsonObjectParser;
+import com.google.api.client.util.GenericData;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.collect.Iterables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ServiceLoader;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Utilities to fetch the S2A (Secure Session Agent) address from the mTLS configuration.
+ *
+ * <p>mTLS configuration is queried from the MDS MTLS Autoconfiguration endpoint.
+ */
+@ThreadSafe
+public final class S2A {
+  public static final String S2A_CONFIG_ENDPOINT_POSTFIX =
+      "/computeMetadata/v1/instance/platform-security/auto-mtls-configuration";
+
+  public static final String METADATA_FLAVOR = "Metadata-Flavor";
+  public static final String GOOGLE = "Google";
+  private static final int MAX_MDS_PING_TRIES = 3;
+  private static final String PARSE_ERROR_S2A = "Error parsing S2A Config from MDS JSON response.";
+
+  private S2AConfig config;
+
+  private transient HttpTransportFactory transportFactory;
+
+  public S2A() {}
+
+  public void setHttpTransportFactory(HttpTransportFactory tf) {
+    this.transportFactory = tf;
+    this.config = getS2AConfigFromMDS();
+  }
```

Comment:
Is there a reason why the config needs to be computed when the HttpTransportFactory is set? To me, it seems a bit odd that setting a value in the setter ends up making a network call. 

What if a user (for whatever reason) ends up calling `setHttpTransportFactory` multiple times?

Reply:
Thanks for pointing this out Lawrence! It is wrong to be setting the config here. Another case where this would be a bug is if the caller doesn't call `setHttpTransportFactory` (if they are using the default instance), in which case the config would never get set. 

In https://github.com/googleapis/google-auth-library-java/pull/1400/commits/8238d50e14e86a529ce0c421079d74b84725a57a I have changed it to set the config in the constructor, and made the transportFactory an optional parameter, rather than having a separate setter. I think this is fine because then the caller can also be aware (without checking the setters) that they are not setting transport and will be using default transport. This also avoids having code like:
```
public String getMtlsS2AAddress() {
    if (config == null) {
      this.config = getS2AConfigFromMDS();
    }
    return config.getMtlsAddress();
  }
public String getPlaintextS2AAddress() {
    if (config == null) {
      this.config = getS2AConfigFromMDS();
    }
    return config.getPlaintextAddress();
  }
```

which introduces a race condition on setting the config.